models/instrument.py

```diff
@@ -1087,6 +1129,148 @@ def get_short_instrument_name(cls):
+           otherwise will throw an exception.
+        existing_ok: bool
+           Only matters if clobber=False (see above)
+        session: models.base.SmartSession
+           Database session to use.  (A new one will be created if this
+           is None, but that will lead to the returned exposures and
+           members of those exposures not being bound to a session, so
+           lazy-loading won't work).
+
+        Returns
+        -------
+        A list of pathlib.Path for the files that were downloaded.
+
+        """
+        raise NotImplementedError( f"Instrument class {self.__class__.__name__} hasn't "
+                                   f"implemented download_exposure." )
+
+    def download_and_load_exposures( self, indexes=None, clobber=False, existing_ok=False,
+                                     delete_downloads=True, skip_existing=True ):
+        """Download exposures and load them into the database.
+
+        Files will first be downloaded to FileOnDiskMixin.local_path
+        with the filename that the origin gave them.  The headers of
+        files will be used to construct Exposure objects.  When each
+        Exposure object is saved, it will copy the file to the file
+        named by Exposure.invent_filpath (relative to
+        FileOnDiskMixin.local_path) and upload the exposure to the
+        archive.
+
+        Parmaeters
+        ----------
+        indexes: list of int or None
+           List of indexes into the set of origin exposures to download;
+           None means download them all.
+        clobber: bool
+           Applies to the originally downloaded file
+           (i.e. FileOnDiskMixin.local_path/{origin_filename}) already
+           exists.  If clobber is True, that originally downloaded file
+           will always be deleted and written over with a redownload.
+           If clobber is False, then if existing_ok is True it will
+           assume that that file is correct, otherwise it throws an
+           exception.
+        existing_ok: bool
+           Applies to the originally downloaded file; see clobber.
+        delete_downloads: bool
+           If True, will delete the originally downloaded files after they have
+           been copied to their final location.  (This mainly exists for testing
+           purposes to avoid repeated downloads.)
+        skip_existing: bool
+           If True, will silently skip loading exposures that already exist in the
+           database.  If False, will raise an exception on an attempt to load
+           an exposure that already exists in the database.
+
+        Returns
+        -------
+        A list of Exposure objects.  Depending on skip_existing, the length of this list may
+        not be the same as the length of indexes.
+
+        """
+        raise NotImplementedError( f"Instrument class {self.__class__.__name__} hasn't "
+                                   f"implemented download_and_load_exposures." )
```

Comment:
There are a few reasons.  First, I don't want to make any general assumptions about the way each instrument stores its internal list of things found.  Every telescope and archive has its own interface, and this is something we're going to need to customize a lot.  (I have used a class hierarchy for this kind of thing in the past -- intermediate subclasses that capture something that's the same for a number of different instruments.)  While a pandas dataframe might be fine for everything, I'm wary of trying to impose structure on things we don't control, i.e. external interfaces for how to get things from telescopes and archives.  It's always surprisingly heterogeneous.

I also don't want to assume that the way you go from whatever's stored internally to figuring out how you copy the image from the server to the local machine is always going to be the same.  It *might* always be a url... but I wouldn't be surprised if sometimes it turned out it wasn't REST-like, but was a URL where you had to pass some parameters in a POST to get the thing you want.

I think we're best off leaving this as explicitly customizable by instrument, rather than trying to anticipate patterns and put them in a general routine.